azurerm/resource_arm_app_service_source_control_token.go

```diff
@@ -0,0 +1,146 @@
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"BitBucket",
+					"Dropbox",
+					"GitHub",
+					"OneDrive",
+				}, false),
+			},
+
+			"token": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Sensitive:    true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"token_secret": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Sensitive:    true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+		},
+	}
+}
+
+func resourceArmAppServiceSourceControlTokenCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).web.BaseClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for App Service Source Control Token creation.")
+
+	scmType := d.Get("type").(string)
+	token := d.Get("token").(string)
+	tokenSecret := d.Get("token_secret").(string)
+
+	locks.ByName(scmType, appServiceSourceControlTokenResourceName)
+	defer locks.UnlockByName(scmType, appServiceSourceControlTokenResourceName)
+
+	properties := web.SourceControl{
+		SourceControlProperties: &web.SourceControlProperties{
+			Token:       utils.String(token),
+			TokenSecret: utils.String(tokenSecret),
+		},
+	}
+
+	if _, err := client.UpdateSourceControl(ctx, scmType, properties); err != nil {
+		return fmt.Errorf("Error updating App Service Source Control Token (Type %q): %s", scmType, err)
+	}
+
+	read, err := client.GetSourceControl(ctx, scmType)
+	if err != nil {
+		return fmt.Errorf("Error retrieving App Service Source Control Token (Type %q): %s", scmType, err)
+	}
+	if read.Name == nil {
+		return fmt.Errorf("Cannot read App Service Source Control Token (Type %q)", scmType)
+	}
+
+	d.SetId(*read.Name)
+
+	return resourceArmAppServiceSourceControlTokenRead(d, meta)
+}
+
+func resourceArmAppServiceSourceControlTokenRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).web.BaseClient
+	ctx := meta.(*ArmClient).StopContext
+
+	scmType := d.Id()
+
+	resp, err := client.GetSourceControl(ctx, scmType)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] App Service Source Control Token (Type %q) was not found - removing from state", scmType)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on App Service Source Control Token (Type %q): %+v", scmType, err)
+	}
+
+	d.Set("type", resp.Name)
+
+	if props := resp.SourceControlProperties; props != nil {
+		d.Set("token", props.Token)
+		d.Set("token_secret", props.TokenSecret)
+	}
+
+	return nil
+}
+
+func resourceArmAppServiceSourceControlTokenDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).web.BaseClient
+	ctx := meta.(*ArmClient).StopContext
+
+	scmType := d.Id()
+
+	// Delete cleans up existing tokens by setting their values to ""
+	token := ""
+	tokenSecret := ""
```

Comment:
out of interest, does this disable this functionality, or just make this password-less? if so, do we want to generate some random (e.g. a GUID) value here instead?

Reply:
Yes, this functionality is "disabled" when token is set to `""`, so it works pretty well for implementing `Delete` :) 

Example: If you first delete the source control token (set token tok ""), then try to deploy an app service with source control and continuous integration (which configures webhooks into online repos like GitHub), you will receive this error from the API:
```
Error: Error updating Source Control for App Service "test-56ea" (Resource Group "test-56ea"): web.AppsClient#CreateOrUpdateSourceControl: Failure sending request: StatusCode=404 -- Original Error: Code="NotFound" Message="Cannot find SourceControlToken with name GitHub." Details=[{"Message":"Cannot find SourceControlToken with name GitHub."},{"Code":"NotFound"},{"ErrorEntity":{"Code":"NotFound","ExtendedCode":"51004","Message":"Cannot find SourceControlToken with name GitHub.","MessageTemplate":"Cannot find {0} with name {1}.","Parameters":["SourceControlToken","GitHub"]}}]
```
If the token is set to anything else than `""` it will try to authenticate with it.